contracts/crosschain/draft-ERC7786Receiver.sol

```diff
@@ -0,0 +1,60 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity ^0.8.0;
+
+import {IERC7786GatewayDestinationPassive, IERC7786Receiver} from "../interfaces/draft-IERC7786.sol";
+
+/**
+ * @dev Base implementation of an ERC-7786 compliant cross-chain message receiver.
+ *
+ * This abstract contract exposes the `receiveMessage` function that is used in both active and passive mode for
+ * communication with (on or multiple) destination gateways. This contract leaves two function unimplemented:
+ *
+ * `_isKnownGateway(address)`, an internal getter used to verify weither an address is recognised by the contract as a
+ * valid ERC-7786 destination gateway. One or multiple gateway can be supported. Note that any malicious address for
+ * which this function returns true would be able to impersonnate any account on any other chain sending any message.
+ *
+ * `_processMessage`, the internal function that will be called with any message that has been validated.
+ */
+abstract contract ERC7786Receiver is IERC7786Receiver {
+    error ERC7786ReceiverInvalidGateway(address gateway);
+
+    /// @inheritdoc IERC7786Receiver
+    function receiveMessage(
+        address gateway,
+        bytes calldata gatewayMessageKey,
+        string calldata source,
+        string calldata sender,
+        bytes calldata payload,
+        bytes[] calldata attributes
+    ) public payable virtual {
+        if (_isKnownGateway(msg.sender)) {
+            // Active mode
+            // no extra check
+        } else if (_isKnownGateway(gateway)) {
+            // Passive mode
+            IERC7786GatewayDestinationPassive(gateway).validateReceivedMessage(
```

Comment:
One answer is that any value should come with a corresponding attribute, and that is how it would be visible from the processMessage (instead of using msg.value)

Reply:
@frangio

Reply:
If we want to make payable compatible with passive mode we would have to add a `value` parameter to `validateReceivedMessage`.

Reply:
After discussion it feels like the thing is:

- Value should only be part of active mode calls.
- Passive mode should not support value, because value cannot be easily moved in a passive way
  - In that case we have a choice of doing `require(msg.value == 0);` in the passive mode, or to allow value to be passed by the relayer, without it being related to the semantics of the message

Reply:
I would do `require(msg.value == 0)`. In terms of the spec I think this check should be recommended but not mandatory.

Reply:
Added a check